src/libs/Network.js

```diff
@@ -25,25 +26,57 @@ let onError = () => {};
 let isOffline;
 Onyx.connect({
     key: ONYXKEYS.NETWORK,
-    callback: val => isOffline = val && val.isOffline,
-});
-
-let didLoadPersistedRequests;
-Onyx.connect({
-    key: ONYXKEYS.NETWORK_REQUEST_QUEUE,
-    callback: (persistedRequests) => {
-        if (didLoadPersistedRequests || !persistedRequests) {
+    callback: (val) => {
+        if (!val) {
             return;
         }
 
-        // Merge the persisted requests with the requests in memory then clear out the queue as we only need to load
-        // this once when the app initializes
-        networkRequestQueue = [...networkRequestQueue, ...persistedRequests];
-        didLoadPersistedRequests = true;
-        Onyx.set(ONYXKEYS.NETWORK_REQUEST_QUEUE, []);
+        // Client becomes online, process the queue.
+        if (isOffline && !val.isOffline) {
+            // Fake merge which trigger the NETWORK_REQUEST_QUEUE Onyx's subscription callback
+            Onyx.merge(ONYXKEYS.NETWORK_REQUEST_QUEUE, []);
+        }
+        isOffline = val.isOffline;
     },
 });
 
+let didLoadPersistedRequests;
+
+ActiveClientManager.isReady().then(() => {
+    Onyx.connect({
+        key: ONYXKEYS.NETWORK_REQUEST_QUEUE,
+        callback: (persistedRequests) => {
+            // NETWORK_REQUEST_QUEUE is shared across clients, thus every client will have similiar copy of
+            // NETWORK_REQUEST_QUEUE. It is very important to only process the queue from leader client
+            // otherwise requests will be duplicated.
+            // We only process the persisted requests when
+            // a) Client is leader.
+            // b) User is online.
+            // c) requests are not already loaded,
+            // d) When there is at least one request
+            if (!ActiveClientManager.isClientTheLeader()
```

Comment:
It kind of feels like to me that `isClientTheLeader()` should be the method that has a promise. It would only resolve once the Active Client Manager has been initialized and is able to make the determination.

Reply:
I am still a big fan of this suggestion based on the following arguments:

- There is no hint anywhere that this method must be run after `ActiveClientManager.isReady()`
- If `isClientTheLeader()` is ever called without `.isReady()` then it's always going to give a false result

This is setting a trap for anyone else that decides to use `isClientTheLeader()` in the future.

Reply:
I am ready to do it but how will I make sure that the value resolved from `isClientTheLeader` is always up-to-date. 

* Why didn't I make isClientLeader a promise?
When we update the ONYXKEYS.NETWORK_REQUEST_QUEUE on ONyx. Onyx.connect is called multiple times due every open browser tab will update the ONYXKEYS.NETWORK_REQUEST_QUEUE. I noticed that If we subscribe to isClientLeader inside this connect callback then an old copy of persistedRequests is passed to the callback which duplicates request.
* Why can't I wrap Onyx.connect call of ONYXKEYS.NETWORK_REQUEST_QUEUE inside isClientLeader a promise which should solve the above problem?
Inside the connect callback we have to check client leadership. If we wrap the connect call inside isClientLeader promise, isClientLeader value will never update after the first resolution. Due to the reason that client leadership change on tab focus. We want to get the updated value here.

Reply:
OK, interesting problem, and makes sense. I think that same problem exists in the current solution?

It almost seems like the promise for `isClientTheLeader()` should be re-resolved each time the ACTIVE_CLIENTS changes. I don't think that's possible though. So, my next theory is that any call to `isClientTheLeader()` should be placed in a callback for ACTIVE_CLIENTS? That seems messy. I'm not sure what to do.

Reply:
IMO, the current implementation is minimally handling the complexities here. The presence of `isReady` does not impose that it is necessary to wrap isClientLeader in this promise which is the actual case. For other places, we don't really need to check the Readiness at all.